enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -0,0 +1,595 @@
+				// TODO: networking
+				Network: nil,
+			},
+			// TODO: grok MaskedPaths and ReadonlyPaths - just copied from podman.
+			MaskedPaths: []string{
+				"/proc/acpi",
+				"/proc/kcore",
+				"/proc/keys",
+				"/proc/latency_stats",
+				"/proc/timer_list",
+				"/proc/timer_stats",
+				"/proc/sched_debug",
+				"/proc/scsi",
+				"/sys/firmware",
+				"/sys/fs/selinux",
+				"/sys/dev/block",
+			},
+			ReadonlyPaths: []string{
+				"/proc/asound",
+				"/proc/bus",
+				"/proc/fs",
+				"/proc/irq",
+				"/proc/sys",
+				"/proc/sysrq-trigger",
+			},
+		},
+	}
+
+	// Provision devices based on host device node info.
+	// TODO: set up devices the same way podman does.
+	for _, path := range []string{
+		"/dev/null",
+		"/dev/zero",
+		"/dev/random",
+		"/dev/urandom",
+	} {
+		d, err := statDevice(path)
+		if err != nil {
+			return nil, fmt.Errorf("get %s device spec: %w", path, err)
+		}
+		spec.Linux.Devices = append(spec.Linux.Devices, *d)
+		spec.Linux.Resources.Devices = append(spec.Linux.Resources.Devices, specs.LinuxDeviceCgroup{
+			Allow:  true,
+			Access: "rw",
+			Type:   d.Type,
+			Major:  &d.Major,
+			Minor:  &d.Minor,
+		})
+	}
+
+	return &spec, nil
+}
+
+func (c *ociContainer) invokeRuntimeSimple(ctx context.Context, args ...string) error {
+	res := c.invokeRuntime(ctx, &repb.Command{}, args...)
+	if res.Error != nil {
+		return res.Error
+	}
+	if res.ExitCode != 0 {
+		if len(res.Stderr) > 0 {
+			return fmt.Errorf("%s", strings.TrimSpace(string(res.Stderr)))
+		} else if len(res.Stdout) > 0 {
+			return fmt.Errorf("%s", strings.TrimSpace(string(res.Stdout)))
+		}
+		return fmt.Errorf("exit code %d", res.ExitCode)
+	}
+	return nil
+}
+
+func (c *ociContainer) invokeRuntime(ctx context.Context, command *repb.Command, args ...string) *interfaces.CommandResult {
+	start := time.Now()
+	defer func() {
+		// DO NOT SUBMIT
+		log.CtxDebugf(ctx, "[%s] %s %s\n", time.Since(start), c.runtime, args[0])
+	}()
+
+	globalArgs := []string{
+		// "strace", "-o", "/tmp/strace.log",
+		c.runtime,
+		"--log-format=json",
+		"--cgroup-manager=cgroupfs",
+	}
+	if *runtimeRoot != "" {
+		globalArgs = append(globalArgs, "--root="+*runtimeRoot)
+	}
+
+	runtimeArgs := append(globalArgs, args...)
+	runtimeArgs = append(runtimeArgs, command.Arguments...)
+
+	// working dir for crun itself doesn't really matter - just default to cwd.
+	wd, err := os.Getwd()
+	if err != nil {
+		return commandutil.ErrorResult(status.UnavailableErrorf("getwd: %s", err))
+	}
+
+	log.CtxDebugf(ctx, "Running %v", runtimeArgs)
+
+	cmd := exec.Command(runtimeArgs[0], runtimeArgs[1:]...)
+	cmd.Dir = wd
+	// TODO: figure out process I/O. crun's process model does not seem to be
+	// compatible with just setting stdout/stderr to an io.Writer like we would
+	// expect - if we set stdout to a bytes.Buffer here for example, then
+	// cmd.Run() will hang. One observation is that crun seems to re-exec
+	// itself, which might have something to do with it.
```

Comment:
Just adding something that I learned the hard way in the past: If `cmd.Stdout` is set to a writer that isn't an `os.File`, Go will spawn a goroutine to copy the output and wait for it in `cmd.Wait()`. This goroutine will not complete until the stdout of the launched process is closed, even if the process itself exited (but e.g. launched a child process that keeps stdout open).

Reply:
Yeah, that behavior has caused problems for me before as well and thought that might have been the issue here too, but I think something different is happening in this case that I don't understand yet. I tried using `os.Pipe()`which returns `os.File` for the read/write ends of the pipe,  and even tried using the lower-level `syscall.ForkExec` to pass the file descriptors for the write-ends of the pipes, but no luck. (Turns out this is effectively what `exec.Command` does under the hood anyway, so I guess it's not too surprising that this didn't work)

I think `crun` does fork a child process, because the `ForkExec` call returns a pid that is different than the `crun` process that I see in `ps aux` while it's hanging. I guess maybe the child process is blocked while trying to write to the pipe, for some reason.

I was briefly looking at https://github.com/containers/conmon which I believe is how podman is spawning / managing the runtime but it will take a bit of time for me to grok.